### PR TITLE
[SYCL] Fix crash during search for undefined functions

### PR DESCRIPTION
**Pull request by Fznamznon**
There was a crash if operator `new` was used inside recursive function
or if undefined overloaded operator `new` was used inside random device
code. The reason for crash is that `FunctionDecl` that represents
operator `new` doesn't have a valid name in AST and `isSYCLUndefinedAllowed`
checks it.

### CODE REVIEW

File: clang/lib/Sema/SemaSYCL.cpp [hunk 1]
```diff
@@ -515,6 +515,10 @@ static bool isSYCLUndefinedAllowed(const FunctionDecl *Callee,
   if (!Callee)
     return false;
 
+  // The check below requires declaration name, make sure we have it.
+  if (!Callee->getIdentifier())
```
Comment (erichkeane) [inline clang/lib/Sema/SemaSYCL.cpp:519]: Note this is also going to include a bunch of the operators too, I doubt you want to restrict those.

Comment (Fznamznon) [reply clang/lib/Sema/SemaSYCL.cpp:519]: Any examples?

Comment (erichkeane) [reply clang/lib/Sema/SemaSYCL.cpp:519]: Any of the operators don't get identifier names.  So something like 'operator+' or 'operator<<', etc.  Even 'operator[]'.

Comment (erichkeane) [reply clang/lib/Sema/SemaSYCL.cpp:519]: Those operators are always FunctionDecls when user declared.  I don't believe they are ever implicitly declared, but they ARE possible to be declared and not defined.

Comment (Fznamznon) [reply clang/lib/Sema/SemaSYCL.cpp:519]: I think we should be fine then. The goal here is to error out when a `FunctionDecl` is declared (but not defined) without `SYCL_EXTERNAL` macro and used inside device code. If a `FunctionDecl` is defined, this code won't even be executed.
If user declared but not defined something without `SYCL_EXTERNAL` macro, and called it, that will be a correct error (by the SYCL spec).
Actually the only purpose of `isSYCLUndefinedAllowed` is allow to call **one** undefined function without `SYCL_EXTERNAL` and this function should have a specific name.

Comment (erichkeane) [reply clang/lib/Sema/SemaSYCL.cpp:519]: >>Actually the only purpose of isSYCLUndefinedAllowed is allow to call one undefined function without SYCL_EXTERNAL and this function should have a specific name.

This is a bit frightning... But I'm just going to back away not-so-slowly, you seem to have a handle of it :)

**Review by elizabethandrews — APPROVED** [review]

**Review by smanna12 — APPROVED** [review]

**Pull request merged**